src/runtime/hexagon/hexagon_user_dma.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <vector>
+
+#ifndef TVM_RUNTIME_HEXAGON_HEXAGON_USER_DMA_H_
+#define TVM_RUNTIME_HEXAGON_HEXAGON_USER_DMA_H_
+
+namespace tvm {
+namespace runtime {
+namespace hexagon {
+
+#define DMA_SUCCESS 0
+#define DMA_FAILURE -1
+
+class HexagonUserDMA {
+ public:
+  /*!
+   * \brief Initiate DMA to copy memory from source to destination address
+   * \param dst Destination address
+   * \param src Source address
+   * \param length Length in bytes to copy
+   * \returns Status, either DMA_SUCCESS or DMA_FAILURE
+   */
+  int Copy(void* dst, void* src, uint32_t length);
+
+  /*!
+   * \brief Wait until the number of DMAs in flight is less than or equal to some maximum
+   * \param max_dmas_in_flight Maximum number of DMAs allowed to be in flight
+   * to satisfy the `Wait` e.g. use `Wait(0)` to wait on "all" outstanding DMAs to complete
+   */
+  void Wait(uint32_t max_dmas_in_flight);
+
+  /*!
+   * \brief Poll the number of DMAs in flight
+   * \returns Number of DMAs in flight
+   */
+  uint32_t Poll();
+
+  //! HexagonUserDMA uses the singleton pattern
+  static HexagonUserDMA& Get() {
+    static HexagonUserDMA* hud = new HexagonUserDMA();
+    return *hud;
+  }
+
+ private:
+  HexagonUserDMA() = default;
+  ~HexagonUserDMA();
+  HexagonUserDMA(const HexagonUserDMA&) = delete;
+  HexagonUserDMA& operator=(const HexagonUserDMA&) = delete;
+  HexagonUserDMA(HexagonUserDMA&&) = delete;
+  HexagonUserDMA& operator=(HexagonUserDMA&&) = delete;
+
+  //! \brief Initializes / resets the Hexagon User DMA engine
+  unsigned int Init();
+
+  //! \brief Calculates the number of DMAs in flight
+  uint32_t DMAsInFlight();
+
+  //! \brief Stores descriptors for all DMAs
+  std::vector<void*> dma_descriptors_;
```

Comment:
Are the descriptors ever deallocated/removed from this list?  If not, it can cause a memory leak.

Reply:
Yes, in the destructor:
```
HexagonUserDMA::~HexagonUserDMA() {
  Init();  // reset DMA engine
  for (auto dma_desc : dma_descriptors_) {
    free(dma_desc);
  }
```

Reply:
Possible issue - @csullivan informs me that the destructor may not get called for Hexagon runtime classes given that we kill the RPC to avoid zombie processes.

Reply:
My concern is that there is only one object of this class, so when the program runs long enough the descriptor allocations will add up.

Reply:
Maybe a good middle ground would be to `free` descriptors once we discover they are marked `done` in the `DMAsInFlight` method.

Reply:
Marking this resolved with the latest commit which will `use ring buffer to store DMA descriptors`